chat-plugins/trivia.js

```diff
@@ -0,0 +1,1180 @@
+if (!triviaData || typeof triviaData !== 'object') triviaData = {};
+if (!triviaData || typeof triviaData.leaderboard !== 'object') triviaData.leaderboard = {};
+if (!Array.isArray(triviaData.ladder)) triviaData.ladder = [];
+if (!Array.isArray(triviaData.questions)) triviaData.questions = [];
+if (!Array.isArray(triviaData.submissions)) triviaData.submissions = [];
+
+const writeTriviaData = (() => {
+	let writing = false;
+	let writePending = false; // whether or not a new write is pending
+
+	let finishWriting = function () {
+		writing = false;
+		if (writePending) {
+			writePending = false;
+			writeTriviaData();
+		}
+	};
+	return function () {
+		if (writing) {
+			writePending = true;
+			return;
+		}
+		writing = true;
+		let data = JSON.stringify(triviaData, null, 2);
+		fs.writeFile('config/chat-plugins/triviadata.json.0', data, () => {
+			// rename is atomic on POSIX, but will throw an error on Windows
+			fs.rename('config/chat-plugins/triviadata.json.0', 'config/chat-plugins/triviadata.json', err => {
+				if (err) {
+					// This should only happen on Windows
+					fs.writeFile('config/chat-plugins/triviadata.json', data, finishWriting);
+					return;
+				}
+				finishWriting();
+			});
+		});
+	};
+})();
+
+// Binary search for the index at which to splice in new questions in a category,
+// or the index at which to slice up to for a category's questions
+function findEndOfCategory(category, inSubmissions) {
+	let questions = inSubmissions ? triviaData.submissions : triviaData.questions;
+	let left = 0;
+	let right = questions.length;
+	let i = 0;
+	let curCategory;
+	while (left < right) {
+		i = (left + right) / 2 | 0;
+		curCategory = questions[i].category;
+		if (curCategory < category) {
+			left = i + 1;
+		} else if (curCategory > category) {
+			right = i;
+		} else {
+			while (++i < questions.length) {
+				if (questions[i].category !== category) break;
+			}
+			return i;
+		}
+	}
+
+	return left;
+}
+
+function sliceCategory(category) {
+	let questions = triviaData.questions;
+	if (!questions.length) return [];
+
+	let sliceUpTo = findEndOfCategory(category, false);
+	if (!sliceUpTo) return [];
+
+	let categories = Object.keys(CATEGORIES);
+	let categoryIdx = categories.indexOf(category);
+	if (!categoryIdx) return questions.slice(0, sliceUpTo);
+
+	// findEndOfCategory for the category prior to the specified one in
+	// alphabetical order returns the index of the first question in it
+	let sliceFrom = findEndOfCategory(categories[categoryIdx - 1], false);
+	if (sliceFrom === sliceUpTo) return [];
+
+	return questions.slice(sliceFrom, sliceUpTo);
+}
+
+class TriviaPlayer extends RoomGamePlayer {
+	constructor(user, game) {
+		super(user, game);
+
+		this.points = 0;
+		this.correctAnswers = 0;
+
+		this.answer = '';
+		this.answeredAt = [];
+		this.isCorrect = false;
+	}
+
+	setAnswer(answer, isCorrect) {
+		this.answer = answer;
+		this.answeredAt = process.hrtime();
```

Comment:
When the script was using milliseconds instead of nanoseconds,  timer mode was sometimes more biased to give you certain numbers of points above others depending on when you answer. Even if the bias would only be 4ms at most,  Ties based off answering time also happen much more rarely now, but were somewhat of a concern before since they happened more often than they should've. I think nanoseconods are fine to use here since the plugin is so sensitive to time that games run a little more predictably tand fairly han they would otherwise.